src/jobserver.h

```diff
@@ -0,0 +1,116 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "graph.h"
+
+/// The GNU jobserver limits parallelism by assigning a token from an external
+/// pool for each command. On posix systems, the pool is a fifo or simple pipe
+/// with N characters. On windows systems, the pool is a semaphore initialized
+/// to N. When a command is finished, the acquired token is released by writing
+/// it back to the fifo or pipe or by increasing the semaphore count.
+///
+/// The jobserver functionality is enabled by passing --jobserver-auth=<val>
+/// (previously --jobserver-fds=<val> in older versions of Make) in the MAKEFLAGS
+/// environment variable and creating the respective file descriptors or objects.
+/// On posix systems, <val> is 'fifo:<name>' or '<read_fd>,<write_fd>' for pipes.
+/// On windows systems, <val> is the name of the semaphore.
+///
+/// The class parses the MAKEFLAGS variable and opens the object handle if needed.
+/// Once enabled, Acquire() must be called to acquire a token from the pool.
+/// If a token is acquired, a new command can be started.
+/// Once the command is completed, Release() must be called to return a token.
+/// The token server does not care in which order a token is received.
+struct Jobserver {
+  /// Return current token count or initialization signal if negative.
+  int Tokens() const { return token_count_; }
+
+  /// Read MAKEFLAGS environment variable and process the jobserver flag value.
+  virtual void Parse() {};
+
+  /// Return true if jobserver functionality is enabled and initialized.
+  virtual bool Enabled() const { return false; }
+
+  /// Implementation-specific method to acquire a token from the external pool
+  /// which is called for all tokens but returns early for the first token.
+  /// This method is called every time Ninja needs to start a command process.
+  /// Return true on success (token acquired), and false on failure (no tokens
+  /// available). First call always succeeds. Ninja is aborted on read errors.
+  virtual unsigned char Acquire() { return '\0'; }
+
+  /// Implementation-specific method to release a token to the external pool
+  /// which is called for all tokens but returns early for the last token.
+  /// Return a previously acquired token to the external token pool.
+  /// It must be called for each successful call to Acquire() after the command
+  /// even if subprocesses fail or in the case of errors causing Ninja to exit.
+  /// Ninja is aborted on write errors, and otherwise calls always succeed.
+  virtual void Release(unsigned char*) {};
```

Comment:
Please explain what the pointer being passed as argument here should point to? E.g. is it ok to call this with a pointer that points to a value that was never the result of a previous Acquire() code (as suggested by your Clear() function implementation). In which case, what is the default value to be used (also apparently `\0` from the code, but you should make that clear in the documentation). Also will the function modify the pointed value or not?

Ideally, users of the API should not have to guess these details by looking at the source code.

I suggest writing a dedicated move-only Token class, even if trivial, to better encapsulate these semantics.

```
/// A wrapper for token values acquired or released to the pool.
/// A default instance has no value, and is used to indicate that
/// no token is available.
struct Token {
  // Default constructor builds a value-less token.
  Token() = default;
  
  // Explicit constructor for a Token with a value from the pipe.
  explicit Token(uint8_t value) : value_(static_cast<int>(value)) {}
  
  // Move operations are allowed.
  Token(Token&& other) noexcept : value_(other.value_) { other.value_ = -1; }
  Token& operator=(Token&&) noexecpt = default;
  
  // Copy operations are forbidden.
  Token(const Token&) = delete;
  Token& operator=(const Token&) noexcept = default;
  
  /// Returns true if this instance contains a value received from the pipe.
  bool HasValue() const { return value_ != -1; }
  
  /// Return underlying value. It is a runtime error to call this method
  /// if HasValue() returns false.
  uint8_t  GetValue() const {
    assert(HasValue());
    return static_cast<uint8_t>(value_); 
  }
  
  int value_ = -1;
};
```

Then you can have `Acquire()`  return a Token by value, and `Release()` take a token by value, and forget about pointers entirely, e.g.:

```
/// Try to acquire a token from the pool. A value-less token instance is returned
/// if no token is available.
virtual Token Acquire() { return Token(); }

/// Release a previously acquire token to the pool. Does nothing if the
/// token argument has no value.
virtual void Release(Token token) {}
```

Reply:
I rewrote comments in the header to reflect the change in functionality and new returns and arguments.

Making a whole class just to manage the value of the tokens seems complicated and overkill to me. It makes the function declarations read better, but the Token class itself is not very readable and adds a lot of object constructs for not much benefit...

Is there any benefit to avoiding pointers here? or avoiding `char`?

What if we just made a simple typedef for the tokens?

I'm very confident that a jobserver will never return a NUL char unless it's to indicate "no tokens available" just as I am using it to mean. I'm also pretty confident that Ninja will never receive anything other than a `'+'` when being used with Make, probably even with forks of Make...